### PR TITLE
Add preliminary back extension support (closes #54)

### LINKED ISSUE #54: Add support for ERA5 back extention and ERA5-Land
ERA5 now has a preliminary back extension (1950-1978).

As it has a seprate entry in the CDS, we need to teach era5cli how to download this and other era5 variants such as ERA5-Land

See here for an overview: https://confluence.ecmwf.int/display/CKB/The+family+of+ERA5+datasets

### PR DESCRIPTION
**Pull request by zklaus**
This adds support for the preliminary back extension. It's not super sophisticated but works for me. It's marked to close #54, but it really only addresses the preliminary back extension, not the era5 land part. However, since the two are quite different in nature, I think it's probably best to add a new issue about that.

Things to check for stylistic compliance:
- order of option in adding to argument parser and `era5cli.fetch.Fetch.__init__`
- docstrings and comments in `era5cli.fetch.Fetch`

Possible modifications/amendments:
- check time range
- (semi-)automatically apply back extension where applicable (years prior to 1979)

### CODE REVIEW

**Comment by zklaus** [comment]
There are three problems according to the checks, the first two of which are unrelated:
- For some reason, the Python 3.5 build doesn't seem to correctly install hdf5. This is likely not related to this PR.
- There is a flake8 due to a short variable name in a list comprehension unrelated to this PR.
- The long name of the option necessitates another line break in the call to the `Fetch` constructor. This creates inserts a newline right after the argument `=`. How do you want to address this?

File: era5cli/cli.py [hunk 1]
```diff
@@ -328,7 +339,9 @@ def _execute(args):
                             statistics=statistics,
                             pressurelevels=args.levels,
                             threads=args.threads,
-                            merge=args.merge)
+                            merge=args.merge,
+                            preliminary_back_extension=
+                            args.preliminary_back_extension)
```
Comment (bouweandela) [inline era5cli/cli.py:344]:
```suggestion
                            args.preliminary_back_extension)
```
You can dedent the arguments, so they are only 1 level indented w.r.t. the function call

**Comment by bouweandela** [comment]
I created a branch `update-tests` to try and solve some of the problems with testing

**Comment by bouweandela** [comment]
> There is a flake8 due to a short variable name in a list comprehension unrelated to this PR.

This is probably due to a newer version of flake8, using a single `l` as a variable name is not very nice because it's easy to mistake an `l` or a `1`.

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@zklaus Thank you, a nice addition to era5cli. Here some issues:

1. product_type: 
I could download the hourly data of the preliminary back extension using this PR.
However, downloading monthly data failed, both for data on pressure level and single level. This is due to the value of `product_type` of monthly average reanalysis: it is `monthly_averaged_reanalysis` in era5 data whereas it is `reanalysis-monthly-means-of-daily-means` for the preliminary back extension. I am not sure if the difference in `product_type` is a bug in cds. It is possible to check `product_type` using "**show API request**" button at the end of these two pages: 
 
- [ERA5 monthly averaged data on single levels from 1979 to present](https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels-monthly-means?tab=form)
- [ERA5 monthly averaged data on single levels from 1950 to 1978 (preliminary version)](https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels-monthly-means-preliminary-back-extension?tab=form).

Therefore, the codes [here](https://github.com/eWaterCycle/era5cli/blob/b1e5a9812829a568cf0bed7322f5d4535aea53b0/era5cli/fetch.py#L210) need a fix accordingly.
 
2. providing the required arguments by users:
Currently, if users provide `preliminary_back_extension`, the `startyear` and `endyear` should be from 1950 to 1978. Could you please add a **raise error statement** if this is not the case, before that the era5cli sends a request to cds. 
 
3. some tests in test_fetch.py:
Could you please add some tests to **test_fetch.py** regarding the new changes?
Thank you.

**Comment by zklaus** [comment]
Thanks for taking over. Sorry I couldn't push it any further. :beers:

**Pull request merged**